### PR TITLE
feat: [CHK-2944] Wallet creation timeout event producer

### PR DESCRIPTION
**Pull request by petretiandrea**
This PR introduce a new functionality to move Wallet to ERROR when it expires. When Wallet moves to `CREATED` state a `WALLET_CREATED` event with visibilityTimeout is produced. `wallet-event-dispatcher` will consume the message and establish next action to do on wallet

#### List of Changes
- added `WalletCreatedEvent` as integration event
- added `DomainEventDispatcher`: a dispatcher which handles produced domain events (e.g. publish an expiration event)
- introduced WalletAddedEvent for Migration to allows to implement different expiration business logic
- improved traceability of queue producer
- minor refactoring

#### How Has This Been Tested?
DEV environment by creating a Wallet and consuming expiration event from `event-dispatcher`

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: helm/values-uat.yaml [hunk 1]
```diff
@@ -167,6 +167,9 @@ microservice-chart:
       { VISA:'https://assets.cdn.platform.pagopa.it/creditcard/visa.png', MASTERCARD:'https://assets.cdn.platform.pagopa.it/creditcard/mastercard.png', DINERS:'https://assets.cdn.platform.pagopa.it/creditcard/diners.png', AMEX:'https://assets.cdn.platform.pagopa.it/creditcard/amex.png', UNKNOWN:'https://assets.cdn.platform.pagopa.it/creditcard/generic.png', PAYPAL:'https://assets.cdn.platform.pagopa.it/apm/paypal.png', MAESTRO: 'https://assets.cdn.platform.pagopa.it/creditcard/maestro.png', MC: 'https://assets.cdn.platform.pagopa.it/creditcard/mastercard.png' }
 
     WALLET_ONBOARDING_PAYPAL_PSP_LIST: "BCITITMM,SIGPITM1XXX"
+    EXPIRATION_QUEUE_NAME: pagopa-u-weu-wallet-expiration-queue
+    EXPIRATION_QUEUE_TTL_SECONDS: "3600"
+    EXPIRATION_QUEUE_TIMEOUT_SECONDS_WALLET_CREATED: "900"
```
Comment (pietro-tota) [inline helm/values-uat.yaml:172]: add those variables to README too :)

Comment (petretiandrea) [reply helm/values-uat.yaml]: You read my mind updated here https://github.com/pagopa/pagopa-wallet-service/pull/148/commits/6f64f337f212802e5051e230de2fb7c5d34f1a68

File: README.md [hunk 2]
```diff
@@ -101,6 +101,10 @@ If you want to customize the application environment, reference this table:
 | REDIS_SSL_ENABLED                                        | Whether SSL is enabled while connecting to  Redis                  | string                   |         |
 | WALLET_SESSION_TTL                                       | Wallet session TTL in minutes                                      | int                      |         |
 | WALLET_LOGO_MAPPING                                      | Wallet logo map parameter that handles wallet type to logo mapping | string (Map<string,URI>) |         |
+| EXPIRATION_QUEUE_NAME                                    | Name of expiration queue                                           | string                   |         |
+| EXPIRATION_QUEUE_TTL_SECONDS                             | TTL in seconds for published message                               | string                   |         |
+| EXPIRATION_QUEUE_CONNECTION_STRING                       | Connection string to storage queue                                 | string                   |         |
+| EXPIRATION_QUEUE_TIMEOUT_SECONDS_WALLET_EXPIRED          | Visibility timeout in seconds for expired event                    |                          |         |
```
Comment (pietro-tota) [inline README.md:107]: what about rename this to `EXPIRATION_QUEUE_VISIBILITY_TIMEOUT_SEC`?

Comment (petretiandrea) [reply README.md]: Solved by https://github.com/pagopa/pagopa-wallet-service/pull/148/commits/dcd30dd8720e5e8f06681487beec27ecc1bf1b43

File: src/test/kotlin/it/pagopa/wallet/common/tracing/TracingUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            .willAnswer {
+                val map: HashMap<String, String> = it.getArgument(1)
+                val setter = it.getArgument<TextMapSetter<HashMap<String, String>>>(2)
+                setter.set(map, TRACEPARENT, "mock_traceparent")
+                null
+            }
+        val contextPropagators = Mockito.mock(ContextPropagators::class.java)
+        given { contextPropagators.textMapPropagator }.willReturn(textMapPropagator)
+        given { openTelemetry.propagators }.willReturn(contextPropagators)
+        Mockito.`when`(openTelemetry.propagators).thenReturn(contextPropagators)
```
Comment (giovanniberti) [inline src/test/kotlin/it/pagopa/wallet/common/tracing/TracingUtilsTest.kt:38]:
```suggestion
        given(openTelemetry.propagators).willReturn(contextPropagators)
```

**Review by giovanniberti — APPROVED** [review]
🚀

File: src/main/kotlin/it/pagopa/wallet/domain/wallets/DomainEventDispatcher.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package it.pagopa.wallet.domain.wallets
+
+import it.pagopa.wallet.audit.LoggingEvent
+import reactor.core.publisher.Mono
+
+interface DomainEventDispatcher {
```
[the remark this answers is not in the document]
Comment (petretiandrea) [reply src/main/kotlin/it/pagopa/wallet/domain/wallets/DomainEventDispatcher.kt:6]: Onestly? I don't known, I will fix it

File: src/test/kotlin/it/pagopa/wallet/common/tracing/TracingUtilsTest.kt (hunk 3 above)
Comment (petretiandrea) [reply src/test/kotlin/it/pagopa/wallet/common/tracing/TracingUtilsTest.kt]: Solved by https://github.com/pagopa/pagopa-wallet-service/pull/148/commits/ee07abeedd9568949c37b35417aa8a6cb2977592

File: src/main/kotlin/it/pagopa/wallet/services/DomainEventDispatcherService.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    private val walletQueueClient: WalletQueueClient,
+    private val tracingUtils: TracingUtils,
+    private val expirationQueueConfig: ExpirationQueueConfig,
+) : DomainEventDispatcher {
+
+    private val logger: Logger = LoggerFactory.getLogger(javaClass)
+
+    private val walletExpireTimeout by lazy {
+        Duration.ofSeconds(expirationQueueConfig.timeoutWalletExpired)
+    }
```
[the remark this answers is not in the document]
Comment (petretiandrea) [reply src/main/kotlin/it/pagopa/wallet/services/DomainEventDispatcherService.kt]: Solved by https://github.com/pagopa/pagopa-wallet-service/pull/148/commits/ee07abeedd9568949c37b35417aa8a6cb2977592

File: src/main/kotlin/it/pagopa/wallet/domain/wallets/DomainEventDispatcher.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (petretiandrea) [reply src/main/kotlin/it/pagopa/wallet/domain/wallets/DomainEventDispatcher.kt]: Solved by https://github.com/pagopa/pagopa-wallet-service/pull/148/commits/ee07abeedd9568949c37b35417aa8a6cb2977592

**Review by giovanniberti — DISMISSED** [review]

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**